src/mds/SessionMap.h

```diff
@@ -661,6 +697,19 @@ class SessionMap : public SessionMapStore {
    */
   void save_if_dirty(const std::set<entity_name_t> &tgt_sessions,
                      MDSGatherBuilder *gather_bld);
+
+private:
+  uint64_t get_session_count_in_state(int state) {
+    return !is_any_state(state) ? 0 : by_state[state]->size();
+  }
+
+  // running counter of session uptime (as nsec)
+  uint64_t total_uptime = 0;
+
```

Comment:
You can add a public method to get the average session **starttime**: keep a running average of the starttimes (the co then divide by the number of sessions when computing the average.

When adding a session:

`avg_starttime = ((avg_starttime / (new_session_count)) * (new_session_count-1) + sesion_starttime/new_session_count`

Removing is the reverse.

The tricky part is doing that with monotonic clock time points rather than unix epochs. @adamemerson, can you advise the right way to compute the average of a set of `coarse_mono_time` epochs? We need to be able to add/remove epochs from the set dynamically, using the math above ^

Reply:
@batrick couldn't it just be average session duration (age)? From the above it seems to me that the average should also take into account ongoing sessions?

Reply:
> @batrick couldn't it just be average session duration (age)?

You can obtain the average session duration via `current_time - average_start_time`?

Reply:
My ultimate point is that you shouldn't need to update the counter in the Session. Age can be determined by the starttime (set once) and the current time (fetched once cheaply).

Reply:
> You can obtain the average session duration via current_time - average_start_time?

I meant having a running counter of `start_time - end_time` average by total number of completed sessions.

Reply:
> My ultimate point is that you shouldn't need to update the counter in the Session. Age can be determined by the starttime (set once) and the current time (fetched once cheaply).

OK -- in that case, I'll revisit the calculation.